[16.0] FastAPI Integration

lmignon opened the pull request:
* [x] manage exceptions and transaction rollback (through the use of a dedicated starlette middleware???)
* [x] manage languages
* [x] works with extendable-pydantic
* [x] depends on https://github.com/OCA/web-api/pull/17
* [x] improves UI to alert the user that the routing map is out of sync and must be synchronized when one element involved into the routing definition is updated or a new app is created. 
* [x] rework the documentation to address the security aspect from the start and to evolve its implementation in the examples given in the following topics.

lmignon committed:
[ADD] Allows use of fastapi into Odoo!

lmignon committed:
[IMP] fastapi: Uses contextvars to hold and provide odoo env on fastapi service excution

lmignon committed:
[IMP] fastapi: renames fastapi.app into fastapi.endpoint

lmignon committed:
[IMP] fastapi: code cleanup

lmignon committed:
[IMP] fastapi: Improves modularity and demo app

* add default empty method to use as dependency to get the authenticated partner
* improves the demo app to illustrate the way the dependency overrides mechanism can be used to provide the right implementation to use to retrieve the authenticated partner according to the security method configured on the app
* add tests for the demo app to show how the TestClient class and the dependey overrides functianality should be used to easily write tests

lmignon committed:
[IMP] fastapi: Add dependency method

This method can be used to get access to the fastapi.endpoint record into your router's methods

lmignon commented:
@Julien00859 Here it's a usage of the `odoo.http.Dispatcher`class to register a new kind of query....

lmignon committed:
[IMP] fastapi: Implement error handling

ensure transation is rolled back in case of error and allows override / extension of the error handling globally or by app

lmignon committed:
[IMP] fastapi: Allows to define the user executing the api calls

lmignon committed:
[IMP] fastapi: Add authenticated_partner_id into ir.rule eval context

lmignon committed:
[IMP] fastapi: Add helper code to ease development of search route handlers

lmignon committed:
[IMP] fastapi: documentation

lmignon committed:
[FIX] add test dependency for fastapi

lmignon commented:
@sebastienbeau @sbidoul @simahawk @StefanRijnhart I took some time to write some documentation explaining what writing a REST api is and how to do it with this new addon. I still have to understand why the tests are failing in our CI but not on my laptop. Your comments and suggestions are welcome.

sbidoul commented on fastapi/depends.py:
```diff
@@ -0,0 +1,107 @@
+    This method is also responsible to put the authenticated partner id
+    into the context of the current environment.
+    """
+    return partner.with_context(authenticated_partner_id=partner.id)
```
Maybe the `with_context` part should be in the `authenticated_partner_env` function?

sbidoul commented on fastapi/depends.py:
```diff
@@ -0,0 +1,107 @@
+def authenticated_partner_from_basic_auth_user(
+    user: Users = Depends(basic_auth_user),  # noqa: B008
+) -> Partner:
+    return user.partner_id
```
Use `env["res.partner"].browse(user.partner_id)`, so the partner is not sudo?

sbidoul commented on fastapi/readme/ROADMAP.rst:
```diff
@@ -0,0 +1,10 @@
+because the integration of the fastapi is based on the use of a specific middleware
+that convert the WSGI request consumed by odoo to a ASGI request. The question
+is to know if it is also possible to develop the same kind of bridge for the
+WebSockets.
```
I suppose the same challenge applies to streaming large responses in the asgi-wsgi bridge?

sbidoul commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1157 @@
+Before you start, we must define some terms:
+
+* **App**: A FastAPI app is a collection of routes, dependencies, and other
+  components that can be used to build a web application.
```
also an ASGI app?

sbidoul commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1157 @@
+
+* **'odoo_env'**: Returns the current odoo environment.
+* **'fastapi_endpoint'**: Returns the current fastapi endpoint model instance.
+* **'authenticated_partner'**: Returns the authenticated partner.
```
Mention authenticated_partner_env and when to use it vs odoo_env?

sbidoul commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1157 @@
+
+    @demo_api_router.get("/partners", response_model=list[PartnerInfo])
+    def get_partners(
+        env=Depends(odoo_env), partner=Depends(authenticated_partner)
```
Does `env` have authenticated_partner_id in its context, or should it be `authenticated_partner_env`?

sbidoul commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1157 @@
+      env: Environment = Depends(odoo_env),  # noqa: B008
+  ) -> Partner:
+      """A dummy implementation that look for a user with the same login
+      as the provided api key
```
Mention this should not be copy-pasted in actual code :)

sbidoul commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1157 @@
+* You can change the implementation of the route handler.
+* You can override the dependencies of the route handler.
+* You can add a new route handler.
+* You can extend the model used as parameter or as response of the route handler.
```
Elaborate this? One can add optional fields (or required fields with default values) to the request.
In the response, one can add fields, or make optional fields required, but not make required fields optional.
One can remove optional fields from a response (depending on what optional means exactly - null vs absent).

sbidoul commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1157 @@
+      _name = "demo.fastapi.endpoint"
+      _description = "Demo Endpoint"
+
+      def echo(self, message: str) -> str:
```
`@api.model`?

sbidoul commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1157 @@
+
+  The liskov substitution principle has also to be respected. That means that
+  if you extend a model, you must add new required fields or you must provide
+  default values for the new optional fields.
```
elaborate?

sbidoul commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1157 @@
+
+* Use plural for the name of a service. For example, if you provide a service
+  that allows you to manage the sale orders, you must use the name 'sale_orders'
+  and not 'sale_order'.
```
Mention the concept of collection of resources which must be named with a plural (`/sale_orders/{id}`?

sbidoul commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1157 @@
+
+
+
+* ... and many more.
```
Say a word about input validation. Be paranoid about what clients send, assume they will send invalid id's, or attempt to access forbidden data.

sbidoul reviewed (commented):
This is seriously great !

sbidoul committed:
Fix a few typos

lmignon committed:
Fix typo

sebastienbeau commented on fastapi/models/fastapi_endpoint_demo.py:
```diff
@@ -0,0 +1,180 @@
+                authenticated_partner_impl_override = (
+                    api_key_based_authenticated_partner_impl
+                )
+        app.dependency_overrides[
+            authenticated_partner_impl
+        ] = authenticated_partner_impl_override
```
```suggestion
            app.dependency_overrides[
                authenticated_partner_impl
            ] = authenticated_partner_impl_override
```

sebastienbeau replied:
Indentation is wrong, so when self.app is different then "demo" it raise an error as authenticated_partner_impl_override is not defined

lmignon committed:
little fixes

flachica commented on fastapi/models/fastapi_endpoint.py:
```diff
@@ -0,0 +1,268 @@
+    def _get_app_exception_handlers(
+        self,
+    ) -> Dict[
+        int | Type[Exception],
```
```suggestion
        Union[int, Type[Exception]],
```

With this notation fastapi support python 3.7+ as you can see [documentation](https://fastapi.tiangolo.com/python-types/#union) section union. Otherwise we would be locked to python 3.10

lmignon replied:
Thank you @flachica 
Happy to see someone making a try with this draft PR. I plan to work on it in the coming days to publish a first release.

flachica replied:
I am thoroughly studying this module and related technology. I intend to use it extensively in the very short term. I understand that it is in an initial state and I will do my best to contribute according to my possibilities and assuming that changes will come and I will have to adapt. Thank you very much for all the effort, especially in the documentation.

lmignon replied:
I've two issues to solve to be complete. 
* ~~The management of the languages to return the translated info into the requested language if available. This will be based on the accept-language http header~~ done in https://github.com/OCA/rest-framework/pull/291/commits/1a750d446939054c5144076b6db3d29858f37218
* ~~The use of extendable-pydantic is not yet completely functional. For some processes, the fastapi lib use internal methods from the pydantic lib without going through the model class directly. This means that the extension mechanisms implemented by the extendable-fastapi library at the class definition level are bypassed in these cases.   I have identified at this stage 2 places that need to be adapted so that the schema for an extended model  is correctly generated in the documentation and the serialization is based on the extended class definition and not the base class.~~ Fixed into the last release of extenable-pydantic (0.0.2)

We'll probably find others issues but I'm confident that nothing should call into question the approach, which seems to me to be validated from a technical point of view. I'll start to use it for my own projects in the coming weeks.

lmignon replied:
@flachica The integration of extendable-pydantic and fastapi is now fixed into the last version of extendable-pydantic (>=0.0.2) I've to adapt the documentation to specify that an extra odoo addon is required to 'extendable_fastapi'. I'll add the management of the languages tomorrow.

lmignon committed:
[FIX] fastapi: Uses new implementation of endpoint_route_handler

lmignon committed:
[FIX] fastapi: change annotation to support python 3.7+

lmignon committed:
[ADD] extendable_fastapi: New addon to allow the use of extendable within fastapi apps

flachica commented on fastapi/__manifest__.py:
```diff
@@ -0,0 +1,25 @@
+        "security/fastapi_endpoint.xml",
+        "views/fastapi_menu.xml",
+        "views/fastapi_endpoint.xml",
+        "views/fastapi_endpoint_demo.xml",
```
```suggestion

```
Posible duplicated resource

flachica commented on fastapi/README.rst:
```diff
@@ -0,0 +1,1294 @@
+    FastAPI Endpoint
+
+Thanks to the **'fastapi.endpoint'** model, you can create as many endpoints as
+you wand and mount as many apps as you want in each endpoint. The endpoint is
```
```suggestion
you want and mount as many apps as you want in each endpoint. The endpoint is
```

Typo

flachica commented on fastapi/README.rst:
```diff
@@ -0,0 +1,1294 @@
+can be used to authenticate a user in odoo. If the authentication is successful,
+we return the partner record linked to the authenticated user.
+
+In some cas you could want to implement a more complex authentication mechanism
```
```suggestion
In some case you could want to implement a more complex authentication mechanism
```
Typo

flachica commented on fastapi/models/fastapi_endpoint_demo.py:
```diff
@@ -0,0 +1,179 @@
+    demo_auth_method = fields.Selection(
+        selection=[("api_key", "Api Key"), ("http_basic", "HTTP Bacic")],
+        string="Authenciation method",
+    )
```
```suggestion
    )
    registry_sync = fields.Boolean(default=True)
```
If the `registry_sync` field is False, the controller is not registered and all URLs return error 404.

With this change I can test the demo contract/API

lmignon replied:
@flachica The motivation to set this field to false by default is to control when the odoo's routing map is reloaded. The reload of the routing map is costly operation and will impact all the odoo runners. I've to improve the UI to highlight that a sync is required. This filed will disappear from the form in favor of something more user friendly.

lmignon committed:
[FIX] fastapi: typo

Co-authored-by: Fernando <fernandolachica@gmail.com>

lmignon committed:
[IMP] extendable_fastapi: Relicence to lgpl

lmignon committed:
[IMP]fastapi: Uses the accept-languages header to determine the expexted processing lang

lmignon committed:
[IMP] fastapi: Register only one handler by application

lmignon committed:
[IMP] extendable_fastapi: Ensure extended pydantic models are working

lmignon committed:
[IMP] fastapi: Improves UX for the endpoint registry synchronisation

A ribbon and a button are dislayed on the form view when a record needs to be synchronized.
In the tree view the records to synchronize are displayed with the text decoration defined for warning. A button is also displayed on the line if sync is required and an server actions allows you to sync selected records in one click

lmignon committed:
[IMP] fastapi: Improves documentation

lmignon commented:
@flachica I think everything is there now. (until proven otherwise :smile: ) Since https://github.com/OCA/web-api/pull/17 is still in draft mode, I keep the same for the current PR. Your review will be welcome on https://github.com/OCA/web-api/pull/17 to speedup the merge.

flachica commented on fastapi/demo/fastapi_endpoint_demo.xml:
```diff
@@ -0,0 +1,54 @@
+        selection_add=[("demo", "Demo Endpoint")], ondelete={"demo": "cascade"}
+    )
+
+    def _get_fastapi_routers(self) -> List[APIRouter]:
```
```suggestion
    def get_fastapi_routers(self) -> List[APIRouter]:
```
IMHO, the [documentation](https://github.com/acsone/rest-framework/blob/16.0-fastapi/fastapi/README.rst?plain=1#L222) is correct and this code must be changed

flachica commented on fastapi/models/fastapi_endpoint.py:
```diff
@@ -0,0 +1,268 @@
+            "dependencies": self._get_fastapi_app_dependencies(),
+        }
+
+    def _get_fastapi_routers(self) -> List[APIRouter]:
```
```suggestion
    def get_fastapi_routers(self) -> List[APIRouter]:
```
IMHO, the [documentation](https://github.com/acsone/rest-framework/blob/16.0-fastapi/fastapi/README.rst?plain=1#L222) is correct and this code must be changed

sbidoul replied:
The documentation was incorrect. This method must not be exposed over xmlrpc.

flachica commented on fastapi/models/fastapi_endpoint_demo.py:
```diff
@@ -0,0 +1,190 @@
+        string="Authenciation method",
+    )
+
+    def _get_fastapi_routers(self) -> List[APIRouter]:
```
```suggestion
    def get_fastapi_routers(self) -> List[APIRouter]:
```
IMHO, the [documentation](https://github.com/acsone/rest-framework/blob/16.0-fastapi/fastapi/README.rst?plain=1#L222) is correct and this code must be changed

flachica commented on fastapi/readme/CONTRIBUTORS.rst:
```diff
@@ -0,0 +1 @@
+* Laurent Mignon <laurent.mignon@acsone.eu>
```
```suggestion
* Laurent Mignon <laurent.mignon@acsone.eu>
* Fernando La Chica <disashop.com>
```

flachica commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1185 @@
+can be used to authenticate a user in odoo. If the authentication is successful,
+we return the partner record linked to the authenticated user.
+
+In some cas you could want to implement a more complex authentication mechanism
```
```suggestion
In some case you could want to implement a more complex authentication mechanism
```
Typo

flachica commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1185 @@
+
+.. code-block:: python
+
+    from fastapi import APIRouter
```
```suggestion
    from fastapi import APIRouter, Depends
```
Depends is used in def get_partners

flachica commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1185 @@
+        email: str
+
+    @demo_api_router.get("/partners", response_model=list[PartnerInfo])
+    def get_partners(env=Depends(odoo_env)) -> list[PartnerInfo]:
```
```suggestion
    async def get_partners(env=Depends(odoo_env)) -> List[PartnerInfo]:
```
In demo code I see that async is added
list -> List is to avoid "TypeError: 'NoneType' object is not callable" exception

flachica commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1185 @@
+    from fastapi import APIRouter
+    from pydantic import BaseModel
+    from odoo import api, fields, models
+    from ..depends import odoo_env
```
```suggestion
    from odoo.addons.fastapi.depends import odoo_env
```
If we build our custom module, depends must be accesed with this notation

flachica replied:
```suggestion
    from ..depends import odoo_env
    from typing import List
```
To avoid "TypeError: 'NoneType' object is not callable" exception

flachica commented on fastapi/fastapi_dispatcher.py:
```diff
@@ -0,0 +1,66 @@
+        # don't parse the httprequest let starlette parse the stream
+        self.request.params = {}  # dict(self.request.get_http_params(), **args)
+        environ = self._get_environ()
+        root_path = "/" + environ["PATH_INFO"].split("/")[1]
```
```suggestion
        root_path = "/".join(environ["PATH_INFO"].split("/")[0:-1])
```
If your endpoint is like `api/v1` the router does not solve the path

flachica commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1185 @@
+        name: str
+        email: str
+
+    @demo_api_router.get("/partners", response_model=list[PartnerInfo])
```
```suggestion
    @demo_api_router.get("/partners", response_model=List[PartnerInfo])
```
To avoid "TypeError: 'NoneType' object is not callable" exception

flachica commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1185 @@
+    from ..depends import odoo_env
+
+    @demo_api_router.get("/partners", response_model=list[PartnerInfo])
+    def get_partners(env=Depends(odoo_env)) -> list[PartnerInfo]:
```
```suggestion
    def get_partners(env=Depends(odoo_env)) -> List[PartnerInfo]:
```
list -> List is to avoid "TypeError: 'NoneType' object is not callable" exception

flachica commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1185 @@
+    @demo_api_router.get("/partners", response_model=list[PartnerInfo])
+    def get_partners(
+        env=Depends(odoo_env), partner=Depends(authenticated_partner)
+    ) -> list[PartnerInfo]:
```
```suggestion
    ) -> List[PartnerInfo]:
```
list -> List is to avoid "TypeError: 'NoneType' object is not callable" exception

flachica commented:
@lmignon Sorry for the number of revisions. I have been contributing as I saw some important issue without accumulating all the suggestions until the end. I hope I have not generated too much noise.

I have already been able to create a service that does Ping/Pong and returns a list of Partners as suggested in the documentation, so we are closer to the goal.

lmignon commented:
@flachica Thank you for your deep review of the documentation and the addon itself. I really appreciate all your comments. It's the first time I take time to write documentation to explain why and how I developed this addon and why and how to use-it. It's true that the provided examples where briefly tested but not in the form they are provided into the documentation. That explain why you have found errors into the code.  Regarding your comments about the use of `sudo`, they are a perfect illustration of an aspect that needs to be improved in this documentation. The `sudo` command is precisely the command not to use when writing the implementation of services. On second thought, I think I need to restructure the documentation to address the security aspects from the beginning and evolve the implementation as the explanations progress, especially once the authentication part is addressed. I'll address this issue in the coming days and keep you informed once it's done.

Regards,

Laurent Mignon

flachica commented:
> @flachica Thank you for your deep review of the documentation and the addon itself. I really appreciate all your comments. It's the first time I take time to write documentation to explain why and how I developed this addon and why and how to use-it. It's true that the provided examples where briefly tested but not in the form they are provided into the documentation. That explain why you have found errors into the code. Regarding your comments about the use of `sudo`, they are a perfect illustration of an aspect that needs to be improved in this documentation. The `sudo` command is precisely the command not to use when writing the implementation of services. On second thought, I think I need to restructure the documentation to address the security aspects from the beginning and evolve the implementation as the explanations progress, especially once the authentication part is addressed. I'll address this issue in the coming days and keep you informed once it's done.
> 
> Regards,
> 
> Laurent Mignon

I really appreciate the effort you have made with the documentation and it has motivated me to use it and improve it according to my possibilities. Also taking advantage of those contributions like sudo that will improve my code.

flachica commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1185 @@
+        """
+        # TODO we should declare a technical user with read access only on the
+        # fastapi.endpoint model
+        return env["fastapi.endpoint"].sudo().browse(_id)
```
```suggestion
        return env["fastapi.endpoint"].browse(_id)
```

flachica replied:
In endpoint sudo not must be used

flachica commented:
I have built my enpoint and it works perfectly. Now it turns out that I've done the tests for that endpoint and it doesn't work. If the enpoint only returns a message it works fine, but if it has to access the database it doesn't throw the exception, nor does it exit self.env[model].search().

I don't know what I'm doing wrong. I'm waiting for the writing of that part of documentation of security that you mention that is pending.

lmignon commented:
@flachica About security you've 2 point to take into account.

1. The user specified on endpoint definition. This user is the user that will be used to run the logic. This user should be put in a specific security group without any access.
2. The authentication mechanism. The authentication mechanism is responsible to provide the partner requesting the partner. In the same time, the system will inject into the evaluation context of the record rules the `authenticated_partner_id` . This will allow you write the required record rules to secure your services and link these rule to the specific security group of your technical user running the services.

The principles are explained here https://github.com/OCA/rest-framework/pull/291/files#diff-d864270e7f736cc705f667b331dd7242c7841de1783c06d10550a27ede6d6ed3R803

flachica commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1185 @@
+    def get_partners(env=Depends(odoo_env)) -> list[PartnerInfo]:
+        return [
+            PartnerInfo(name=partner.name, email=partner.email)
+            for partner in env["res.partner"].search([])
```
This method cannot be tested. When running `search` the system crashes. The endpoint cannot access the database during the test. However, it can be executed from `Postman` e.g.

lmignon committed:
[IMP] fastapi: Improves documentation on security aspects

In the same time, applies the security guidelines for the demo app

lmignon commented:
@flachica In my last commit I improved the documentation about security aspects and provides now a basic group to use when creating your own security group for your app.

flachica commented:
> @flachica In my last commit I improved the documentation about security aspects and provides now a basic group to use when creating your own security group for your app.

I have read it. It's great

guewen commented on fastapi/models/fastapi_endpoint_demo.py:
```diff
@@ -0,0 +1,190 @@
+        selection_add=[("demo", "Demo Endpoint")], ondelete={"demo": "cascade"}
+    )
+    demo_auth_method = fields.Selection(
+        selection=[("api_key", "Api Key"), ("http_basic", "HTTP Bacic")],
```
```suggestion
        selection=[("api_key", "Api Key"), ("http_basic", "HTTP Basic")],
```

florian-dacosta commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1240 @@
+  <record id="my_demo_app_group" model="res.groups">
+    <field name="name">My Demo Endpoint Group</field>
+    <field name="users" eval="[(4, ref('my_demo_app_user'))]" />
+    <field name="implied_ids" eval="[(4, ref('fast_api.group_fastapi_endpoint_runner'))]" />
```
fast_api -> fastapi

jbaudoux commented on fastapi/README.rst:
```diff
@@ -0,0 +1,1377 @@
+
+.. code-block:: python
+
+    from fastapi import APIRouter
```
```suggestion
    from fastapi import APIRouter, Depends
```

jbaudoux commented on fastapi/README.rst:
```diff
@@ -0,0 +1,1377 @@
+    from fastapi import APIRouter
+    from pydantic import BaseModel
+    from odoo import api, fields, models
+    from ..depends import odoo_env
```
```suggestion
    from odoo.addons.fastapi.depends import odoo_env
```

jbaudoux commented on fastapi/README.rst:
```diff
@@ -0,0 +1,1377 @@
+
+.. code-block:: python
+
+    from ..depends import odoo_env
```
```suggestion
    from odoo.addons.fastapi.depends import odoo_env
```

jbaudoux commented on fastapi/readme/USAGE.rst:
```diff
@@ -0,0 +1,1240 @@
+
+.. code-block:: python
+
+    from ..depends import odoo_env
```
```suggestion
    from odoo.addons.fastapi.depends import odoo_env
```

jbaudoux replied:
```suggestion
    from odoo.addons.fastapi.depends import odoo_env
```

linhhonblade commented:
May I know the fastapi version you used? I tried to run these code recently but got error "cannot import name 'APIRouter' from 'fastapi'"

lmignon committed:
[IMP] fastapi: Declared as Beta

lmignon committed:
[IMP] fastapi: Typos

lmignon commented:
> May I know the fastapi version you used? I tried to run these code recently but got error "cannot import name 'APIRouter' from 'fastapi'"

AFAIK Tha APIRouter is in all versions of fastapi https://github.com/tiangolo/fastapi/blob/c81e136d75f5ac4252df740b35551cf2afb4c7f1/fastapi/routing.py#L478

flachica commented:
Dear @lmignon, while reviewing the security issue of the `fastapi` module, I realized that it may be necessary to adapt the [base_rest_auth_user_service](https://github.com/OCA/rest-framework/tree/16.0/base_rest_auth_user_service) module to depend on it and not on `base_rest`. (And make it work with Odoo OAuth if it is possible)

The tests I have done have been from `Postman` and skipping the need to provide security to the system. But that time has passed. 

The question is this: Is there any plan to move the base_rest dependent modules that are already in 16.0 to fastapi? Is there any progress in this regard? It is not to duplicate work. 

I would like to contribute and I think that in order to maximize the value it is good to reach an agreement.

lmignon committed:
Fix schema generation

When a field is of an extended type, the assemebled class is not part of the know classes computed by fastapi from the fastapi endepoint definition. Is such a case, we must add the extented type to the list of knwon classes

lmignon committed:
[FIX] fastapi: odoo-addon-endpoint-route-handler is merged

lmignon commented:
/ocabot merge nobump

sbidoul commented on fastapi/depends.py:
```diff
@@ -0,0 +1,123 @@
+    to declare the way your partner will be provided. In some case, this
+    partner will come from the authentication mechanism (ex jwt token) in other cases
+    it could comme from a lookup on an email received into an HTTP header ...
+    See the fastapi_endpoint_demo for an example"""
```
@lmignon should this not raise `NotImplementedError` ?

sbidoul replied:
Ok, I'll do that in the `fastapi_auth_jwt` PR.

lmignon commented:
Yes, it should be.

On Thu, Jun 8, 2023 at 12:30 PM Stéphane Bidoul ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In fastapi/depends.py
> <https://github.com/OCA/rest-framework/pull/291#discussion_r1222837897>:
>
> > +from .schemas import Paging
> +
> +if TYPE_CHECKING:
> +    from .models.fastapi_endpoint import FastapiEndpoint
> +
> +
> +def odoo_env() -> Environment:
> +    yield odoo_env_ctx.get()
> +
> +
> +def authenticated_partner_impl() -> Partner:
> +    """This method has to be overriden when you create your fastapi app
> +    to declare the way your partner will be provided. In some case, this
> +    partner will come from the authentication mechanism (ex jwt token) in other cases
> +    it could comme from a lookup on an email received into an HTTP header ...
> +    See the fastapi_endpoint_demo for an example"""
>
> @lmignon <https://github.com/lmignon> should this not raise
> NotImplementedError ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OCA/rest-framework/pull/291#pullrequestreview-1469587593>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEE2WXPP2XWDBZ4TQ2YECTXKGSSVANCNFSM6AAAAAAQXOT7K4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Josesosa07 commented:
Hi, I am using  in the manifest: ["base", "base_rest_auth_api_key", "base_rest_datamodel",] 
but I got a warning saying: rest-framework/base_rest/__init__.py: base_rest is deprecated and not fully supported anymore on Odoo 16. Please migrate to the FastAPI migration module. See https://github.com/OCA/rest-framework/pull/291.
I am in a project in odoo 16, if I continue using the rest-api in this way, will I get issues in the project?

Do I need to use fastapi? If so, how should it be implemented?

sbidoul commented:
@Josesosa07 `base_rest` is deemed deprecated because it's main maintainers (@lmignon and ACSONE) have decided to move towards a FastAPI integration which is much more powerful and simplier, delegating a lot of the heavy lifting to the fastapi library.

So if you are starting a new project, it is better to start with the `fastapi` addon in this repo, which also has extensive documentation on how to get started. If you are porting existing `base_rest` services from a previous version and don't want to convert them to FastAPI, `base_rest` is still just fine.
